### PR TITLE
Add users spam detector task

### PR DESCRIPTION
**Pull request by laurajaime**
## Add users spam detector task

This rake task export a .csv with a list of suspicious users.

- `bin/rake cdtb:spam:users[org_id]`

Examples:
`bin/rake cdtb:spam:users[1]` --> find users in organization with id=1.
`bin/rake cdtb:spam:users` --> find all users in all organizations.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -76,6 +76,19 @@ To migrate from S3 to local storage, the identified steps will be:
...
+
+To detect spam in Decidim.
+
+#### Detect spam users
+This rake task export a .csv with a list of suspicious users.
+
+- `bin/rake cdtb:spam:users[host]`
+
+Examples:
+`bin/rake cdtb:spam:users[boi]` --> find users in organization with "boi" string in the host name.
```
Comment (tramuntanal) [inline README.md:89]: what will happen if 5 organizations match? I think it is better to pass the organization ID

File: lib/decidim/cdtb/spam/user_spam_detector.rb [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+module Decidim
+  module Cdtb
+    module Spam
+      # Detect spam behavior in users
+      #
+      class UserSpamDetector < ::Decidim::Cdtb::Task
+        DEFAULT_SPAM_WORDS = %w[viagra sex game free crypto crack xxx luck girls vip download].freeze
+
+        def initialize(host = nil)
```
Comment (tramuntanal) [inline lib/decidim/cdtb/spam/user_spam_detector.rb:13]: receive the organization or nil

File: lib/decidim/cdtb/spam/user_spam_detector.rb [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+          super("SPAM DETECTOR", progress_bar: progress_bar)
+        end
+
+        def prepare_execution(_ctx)
+          if @host.present?
+            @query = Decidim::Organization.where("host ilike ?", "%#{@host}%")
+            @users = Decidim::User.where(organization: @query.first)
+          else
+            @users = Decidim::User.all
+          end
```
Comment (tramuntanal) [inline lib/decidim/cdtb/spam/user_spam_detector.rb:25]:
```suggestion
          @query = Decidim::Organization
          @query = @query.where("host ilike ?", "%#{@host}%") if @organization
          @users = @query.all
```

File: lib/decidim/cdtb/spam/user_spam_detector.rb [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+              @num_applied+= 1
+            end
+
+            progress_bar.increment
+          end
+
+          if @suspicious_users.present?
+            export_users_to_csv
+          else
+            log_task_end
```
Comment (tramuntanal) [inline lib/decidim/cdtb/spam/user_spam_detector.rb:51]: this is automatically done

File: lib/decidim/cdtb/spam/user_spam_detector.rb [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+          @num_users
+        end
+
+        def do_execution(context)
+          progress_bar = context[:progress_bar]
+          @suspicious_users = []
+
+          @users.find_each do |user|
+            if spam_user?(user)
+              @suspicious_users << user
```
Comment (tramuntanal) [inline lib/decidim/cdtb/spam/user_spam_detector.rb:41]: There are organizations with thousands of spam users. The system resources may collapse if we add so many User objects into memory.

The CSV must be generated on the fly during the find_each :pray:

File: lib/decidim/cdtb/spam/user_spam_detector.rb [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+            @users.each do |user|
+              suspicious = suspicious_emails.include?(user.email) ? "✅" : "❌"
+              csv << [user.name, user.email, user.nickname, suspicious]
+            end
+          end
+        end
+
+        def has_spam_word?(user)
+          DEFAULT_SPAM_WORDS.any? do |word|
```
Comment (tramuntanal) [inline lib/decidim/cdtb/spam/user_spam_detector.rb:86]: Those defaults must go into a new initializer for this module, and should be overwritable via another application initializer.This is:
- this module should have an initializer that sets these default values
- in the application where this module is installed, there should be the option (it is optional) to add an initializer to overwrite the default values.

**Review by tramuntanal — CHANGES_REQUESTED** [review]
Good job on the first version @laurajaime !
but we need to rework it a little bit.

Please, read my comments and if you have any question let's comment.

File: lib/decidim/cdtb/spam/user_spam_detector.rb (hunk 3 above)
Comment (laurajaime) [reply lib/decidim/cdtb/spam/user_spam_detector.rb:25]: But here `@query.all` is organizations, not users like in my code.

**Comment by laurajaime** [comment]
@tramuntanal rubocop is failing on the length of the rspec describe. Should I increase this length in the rubocop.yml? what do you think?

**Comment by tramuntanal** [comment]
> rubocop is failing on the length of the rspec describe

ok, increase to 50 for example

File: lib/tasks/spam.rake [hunk 7]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+namespace :cdtb do
+  namespace :spam do
+    desc "Show a list with users suspected of spam"
+    task :users, %i[org_id] => :environment do |_task, args|
+      organization = args.org_id.present? ? Decidim::Organization.find(args.org_id) : Decidim::Organization.all
```
Comment (tramuntanal) [inline lib/tasks/spam.rake:7]:
```suggestion
      organization = args.org_id.present? ? Decidim::Organization.find(args.org_id) : nil
```

File: lib/decidim/cdtb/spam/user_spam_detector.rb [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+module Decidim
+  module Cdtb
+    module Spam
+      # Detect spam behavior in users
+      #
+      class UserSpamDetector < ::Decidim::Cdtb::Task
+        include ActiveSupport::Configurable
+
+        URL_REGEX = %r{(http|ftp|https)://([\w_-]+(?:(?:\.[\w_-]+)+))([\w.,@?^=%&:/~+#-]*[\w@?^=%&/~+#-])}.freeze
```
Comment (tramuntanal) [inline lib/decidim/cdtb/spam/user_spam_detector.rb:13]: Still not working, even escaping double slashes: `(http|ftp|https):\/\/([\w_-]+(?:(?:\.[\w_-]+)+))([\w.,@?^=%&:/~+#-]*[\w@?^=%&/~+#-])`  see https://rubular.com/r/iLWUgZrwb35pBb

File: config/initializers/spam_words.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+Decidim::Cdtb::Spam::UserSpamDetector.configure do |config|
+  config.spam_words = %w[viagra sex game free crypto crack xxx luck girls vip download]
+
+  config.spam_regexp = Regexp.union(config.spam_words)
+end
```
Comment (tramuntanal) [inline config/initializers/spam_words.rb:7]: This does not allow what I meant in my previous comment. How would you overwrite the spam words from the app? repeating this same initializer will expose the internals of the module.

This should only be a `Decidim::Cdtb.configure` with one single value, the `spam_words`. Then we must manage to generate the regexp depending on the value of the `spam_words`. I realize that we even can skip the initializer in the module and have an internal default value in the configuration.

File: lib/decidim/cdtb/spam/user_spam_detector.rb [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+                suspicious = "✅"
+                @num_applied+= 1
+              end
+
+              csv << [user.name, user.email, user.nickname, suspicious]
+            end
+          end
+
+          progress_bar.increment
+        end
```
[the remark this answers is not in the document]
Comment (tramuntanal) [reply lib/decidim/cdtb/spam/user_spam_detector.rb]: This change is pending

File: lib/decidim/cdtb/spam/user_spam_detector.rb [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+            csv_headers = ["ID", "Is suspicious?", "Name", "Email", "Nickname", "Personal URL", "About"]
+            csv << csv_headers
+
+            @users.find_each do |user|
+              suspicious = "❌"
+
+              if spam_user?(user)
+                suspicious = "✅"
+                @num_applied+= 1
+              end
```
Comment (tramuntanal) [inline lib/decidim/cdtb/spam/user_spam_detector.rb:51]: Avoid exporting emojis, better nil/x or TRUE/FALSE or yes/not i don't know the standard way of exporting a boolean
```suggestion
              suspicious = false

              if spam_user?(user)
                suspicious = true
                @num_applied+= 1
              end
```

**Review by tramuntanal — CHANGES_REQUESTED** [review]
Still some rework to do :pray:

File: lib/decidim/cdtb/spam/user_spam_detector.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (laurajaime) [reply lib/decidim/cdtb/spam/user_spam_detector.rb]: I don't understand... what change must I do there?

File: lib/decidim/cdtb/spam/user_spam_detector.rb (hunk 8 above)
Comment (laurajaime) [reply lib/decidim/cdtb/spam/user_spam_detector.rb:13]:
```
irb(main):001:0> URL_REGEX = %r{(http|ftp|https)://([\w_-]+(?:(?:\.[\w_-]+)+))([\w.,@?^=%&:/~+#-]*[\w@?^=%&/~+#-])}.freeze
=> /(http|ftp|https):\/\/([\w_-]+(?:(?:\.[\w_-]+)+))([\w.,@?^=%&:\/~+#-]*[\w@?^=%&\/~+#-])/
irb(main):002:0> email = "user@example.com"
=> "user@example.com"
irb(main):003:0> email =~ URL_REGEX
=> nil
irb(main):004:0> !!(email =~ URL_REGEX)
=> false
irb(main):005:0> url = "http://example.org"
=> "http://example.org"
irb(main):006:0> !!(url =~ URL_REGEX)
=> true
```

File: config/initializers/spam_words.rb (hunk 9 above)
Comment (tramuntanal) [reply config/initializers/spam_words.rb:7]: In the application an initializer:

```
Decidim::Cdtb.spam_words= ENV['CDTB_SPAM_WORDS']&.split(",")
```
or

```
Decidim::Cdtb.configure do |config|
  config.spam_words= ENV['CDTB_SPAM_WORDS']&.split(",")
end
```

Comment (tramuntanal) [reply config/initializers/spam_words.rb:7]: Then in this module's engine:
```
config.after_initialize do
  Decidim::Cdtb.config.spam_regexp = Regexp.union(Decidim::Cdtb.config.spam_words)
end
```

File: lib/decidim/cdtb/spam/user_spam_detector.rb (hunk 8 above)
Comment (tramuntanal) [reply lib/decidim/cdtb/spam/user_spam_detector.rb:11]: still not working... https://rubular.com/r/jIE5CWSnjJKhdX

File: README.md [hunk 12]
```diff
@@ -76,6 +76,26 @@ To migrate from S3 to local storage, the identified steps will be:
...
+
+Examples:
+`bin/rake cdtb:spam:users[org_id]` --> find users in organization with an id.
+`bin/rake cdtb:spam:users` --> find all users in all organizations.
+
+To set custom words in the rake, you can override it with an initalizer:
+
+Decidim::Cdtb.configure do |config|
+  config.spam_words = ENV["CDTB_SPAM_WORDS"]&.split(",")
+end
```
Comment (tramuntanal) [inline README.md:97]: Good! le'ts make it more readable
````suggestion
```
Decidim::Cdtb.configure do |config|
  config.spam_words = ENV["CDTB_SPAM_WORDS"]&.split(",")
end
```
````

File: lib/decidim/cdtb/spam/user_spam_detector.rb (hunk 8 above)
Comment (tramuntanal) [reply lib/decidim/cdtb/spam/user_spam_detector.rb:11]: yo can try with: https://rubular.com/r/6EaLDcTjBZHd95

**Review by tramuntanal — COMMENTED** [review]
It is nearly done!
You must check the URL_REGEX

File: README.md (hunk 12 above)
Comment (laurajaime) [reply README.md:97]: the suggestion is the same

File: lib/decidim/cdtb/spam/user_spam_detector.rb (hunk 8 above)
Comment (tramuntanal) [reply lib/decidim/cdtb/spam/user_spam_detector.rb]: Good!!!! :partying_face:

**Review by tramuntanal — APPROVED** [review]
Good, thanks for the patience!

**Pull request merged**